Refector: Coordinator Protocol 변경

Jeon0976 opened the pull request:
## Motivation ⍰

- 기존 순환참조 되는 문제를 찾고보니 main을 갖고있는 NavigationController의 ViewController remove를 하지 않았기 때문임. 
- TabBar, Login이랑은 다른 NavigationContrller 
- 그러므로 각 NavigationController를 직접 Remove할 수 있도록 extension으로 옮김 
- extension에서는 
- 만약 childCoordinator가 존재하면 해당 coordinator에서 navigaionController.ViewController.removeAll()를 수행 후 coordinator를 제거 

<br>

## Key Changes 🔑

-  coordinator extension 변경 
- tabbarCoordinator에서 TabBarViewContrller weak var 처리 

<br>

## To Reviewers 🙏🏻

- 

<br>

## Linked Issue 🔗

-

Jeon0976 committed:
Refector: Coordinator Protocol 변경

Minny27 commented on Nagaza/Sources/Presentation/Feature/Flow/TabBarFlowCoordinator.swift:
```diff
@@ -17,28 +17,29 @@ final class TabBarFlowCoordinator: Coordinator {
     weak var tabBarDelegate: TabBarDelegate? = nil
     
-    private var tabBarController: NagazaTabBarController
+    private weak var tabBarVC: NagazaTabBarController!
```
탭 바에 대한 참조가 private하게 쓰여서 weak는 따로 없어도 될 거 같습니다.

Jeon0976 replied:
tabBar, login, main 등 app에 있는 NavigationController를 제외하고는 상위 프로토콜에서 주입하면서 상위 프로토콜에서 제거 해주기 때문에 순환참조 방지로 weak 처리 해줬습니다. 나머지는 전부 weak했는데 tabbar만 안해서 weak 처리 해줬습니다

Jeon0976 replied:
Login, Main, Map 등 등 다른 ViewController에서 push등 행위를 할때 flow내부에 viewcontroller가 있나 없나 체크를 위해 남겨뒀는데 그거에 맞춰서 tabBar도 변경했는데 tabBar내부에서는 따로 push 등 행위를 할 필요가 없으니, 윌님 말씀처럼 weak를 안 해도 될것같네용

Minny27 commented on Nagaza/Sources/Presentation/Utils/Common/Coordinator.swift:
```diff
@@ -22,6 +22,13 @@ protocol Coordinator: AnyObject {
 extension Coordinator {
     func finish() {
+        
+        if !childCoordinators.isEmpty {
```
isEmpty 조건은 확인 안 해도 될 거 같습니다

Jeon0976 replied:
logout을 하는 flow에서 finish하고 tabBar flow에서 finish할 때 logout하는 Flow는 child가 없을 수 있으니 조건 처리 했는데 필요없을것같다면 조건 처리는 빼는방향으로 갈꼐요

Minny27 replied:
아 !isEmpty을 쓸 쓰나 안 쓰나 동일한 로직인 거 같아서 뻈습니다

Minny27 committed:
[#3]Refactor: weak 삭제

- tab bar vc는 private하게 사용되서 weak 참조 선언하지 않아도 될 거 같습니다
- childCoordinators.isEmpty 조건은 빼도 될 거 같습니다

Minny27 commented:
Coordinator 프로토콜에 있는 navigationController이나 childCoordinators도 weak 참조로 선언해야하는 건 아닌가에 대한 생각이 들긴하네요. 이거는 잘 모르겠네요

Minny27 commented:
고생하셨습니다!
